.github/scripts/abtest.py

```diff
@@ -0,0 +1,159 @@
+    if os.path.exists(version_py_path):
+        os.remove(version_py_path)
+    try:
+        # some packages are not included in the wheel, so use `develop`, not `install`
+        command = ["python", "setup.py", "develop"]
+        # setup environment variables
+        build_env = setup_build_env(os.environ.copy())
+        subprocess.check_call(command, cwd=repo_path, env=build_env, shell=False)
+        command_testbuild = ["python", "-c", "'import torch'"]
+        subprocess.check_call(command_testbuild, cwd=os.environ["HOME"], env=build_env, shell=False)
+    except subprocess.CalledProcessError:
+        # If failed, remove the build directory, then try again
+        build_path = os.path.join(repo_path, "build")
+        if os.path.exists(build_path):
+            shutil.rmtree(build_path)
+        subprocess.check_call(command, cwd=repo_path, env=build_env, shell=False)
+    print("done")
+
+def process_test_result(result_a: Path, result_b: Path, output_dir: str) -> str:
+    def validate_results(a, b) -> bool:
+        metrics = a["metrics"].keys()
+        return sorted(metrics) == sorted(b["metrics"])
+    # check two results are different files
+    assert not result_a == result_b, f"Path {result_a} and {result_b} are the same. Exit."
+    # validate results
+    with open(result_a, "r") as fa:
+        a = json.load(fa)
+    with open(result_b, "r") as fb:
+        b = json.load(fb)
+    assert validate_results(a, b), f"Result validation failed for {result_a} and {result_b}."
+    # print result in csv format
+    header = ["Metric", a["environ"]["pytorch_git_version"][:GIT_HASH_LEN], b["environ"]["pytorch_git_version"][:GIT_HASH_LEN]]
+    out = [header]
+    metrics = a["metrics"].keys()
+    for m in sorted(metrics):
+        val = [m]
+        val.append(a["metrics"][m])
+        val.append(b["metrics"][m])
+        out.append(val)
+    out = "\n".join([";".join(map(lambda y: str(y), x)) for x in out])
+    os.makedirs(output_dir, exist_ok=True)
+    with open(os.path.join(output_dir, "control.json"), "w") as fcontrol:
+        json.dump(a, fcontrol, indent=4)
+    with open(os.path.join(output_dir, "treatment.json"), "w") as ftreatment:
+        json.dump(b, ftreatment, indent=4)
+    with open(os.path.join(output_dir, "result.csv"), "w") as fout:
+        fout.write(out + "\n")
+    return out
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--pytorch-repo", required=True, type=str, help="PyTorch repo path")
+    parser.add_argument("--base", required=True, type=str, help="PyTorch base commit")
+    parser.add_argument("--head", required=True, type=str, help="PyTorch head commit")
+    parser.add_argument("--userbenchmark", required=True, type=str, help="Name of the userbenchmark to run")
+    parser.add_argument("--output-dir", required=True, type=str, help="Output dir path")
+    parser.add_argument("--skip-build", action="store_true", help="Skip PyTorch build")
```

Comment:
What's the use case for skip_build? It looks like if `skip_build=True` then `result_a` and `result_b` would be the same. Maybe it should be do skip_build separately for base or head for the case where base or head is already built?

Reply:
I mainly find `skip_build` is useful in debugging or A-A testing, for example, when the userbenchmark output json is strange, or we want to know if the userbenchmark gives stable result. It saves the time of building pytorch from scratch.